### PR TITLE
perf: COUNT/aggregate fast paths — close the SPARQL aggregate benchmark gaps + overlay support

### PR DESCRIPTION
**Pull request by bplatz**
Based on #1265 

## Summary

Overhauls `COUNT` / `SUM` / aggregate execution with metadata-driven fast paths, asymmetric-seek joins, and parallel partitioned merges, collapsing the slow-spot aggregate queries on the DBLP "KG with associated data" benchmark (1.574B triples, r7a.4xlarge / 16-core). It then extends the **entire** count/filter family to fold uncommitted **novelty** and honor **time-travel** (parallel-under-overlay + a novelty-delta path), and lands the class-stat correctness fixes that make the metadata folds exact on incremental indexes — not just bulk imports.

## Results (DBLP, as of `c32b5cd8`)

Per-query latency on the targeted slow-spots dropped by up to ~18×:

| query | before (ms) | after (ms) | speedup |
|---|---|---|---|
| filter-many-results | 6,576 | 375 | ~18× |
| minus-join-3-star-2 | 4,725 | 329 | ~14× |
| union-constraint-large-join | 5,129 | 373 | ~14× |
| exists-join-3-star-2 | 4,673 | 330 | ~14× |
| optional-join-large-large | 4,455 | 372 | ~12× |
| filter-few-results | 2,761 | 370 | ~7× |

*(The overlay-parallel lanes and the bound-class fold landed after this run and were not re-benchmarked — they target live-write latency, not the bulk-imported benchmark, which is queried at HEAD.)*

### 1. HEAD COUNT/aggregate fast paths (the benchmark wins)
- **Metadata folds (zero scan):** single-predicate `COUNT(*)` and `COUNT(DISTINCT ?p)` from branch-manifest / leaflet counts; no-constraint `UNION` = Σ per-branch; rdf:type inner-star `COUNT(*)` from per-(class,property) stats; POST numeric-compare global min/max shortcut; **bound-class** property `COUNT(*)` (`?s a :C . ?s P ?o` → `classStat[:C][P]`).
- **Asymmetric seek** for skewed small+large joins: drive from the small predicate and leapfrog-seek the large one (`PsotSubjectSeek`) — join / OPTIONAL / MINUS / EXISTS.
- **Parallel partitioned merge** for both-large cases: partition the subject space at leaf boundaries and run the N-way merge per range across cores — inner-star, OPTIONAL, constrained-UNION, MINUS/EXISTS.
- **Parallel filter-scan:** `COUNT` with encoded FILTERs (`?s != ?o`, `isBlank`, `LANG`) and `SUM(?o cmp K)` / numeric-compare `COUNT`, split across leaf chunks.
- **GROUP BY ?o COUNT ORDER BY DESC LIMIT:** streaming run-length + bounded top-K heap.

### 2. Shared thread pool (multi-tenant)
Parallel count/scan work runs on the shared global thread pool instead of a per-query `thread::scope`, so N concurrent queries no longer each spawn a fresh K-thread fan-out — total worker threads stay bounded to ≈ cores.

### 3. Overlay / time-travel support
The whole count/filter family now folds uncommitted novelty and honors `to_t`:
- **Parallel-under-overlay:** per-partition bounded overlay cursors (parallel base scan + per-partition novelty merge via the proven `merge_overlay_into_batch`) — filter-scans and all four count-plan reducers (star / optional / union / minus-exists).
- **Novelty-delta:** single-predicate and no-constraint `UNION` `COUNT(*)` keep the instant manifest base count and rescan only the leaves novelty touches (`base_total − base(touched) + merged(touched)`).
- Gated to `to_t >= max_t` (at/above the indexed head); time-travel below the indexed head defers to the replay path.

### 4. Class-stat correctness (enables the folds on any index)
- **#1266:** incremental per-(class,property) datatype/language stats made current-state-exact (base-vs-net class attribution + base-index re-scan of re-typed subjects) → the rdf:type-star fold drops its bulk-import-only gate and runs on any base index.
- **Ref-class edge re-attribution on re-type** (outgoing edges from the same base scan + inbound edges via a reverse OPST lookup) plus a **rebuild gate** that falls back to a full rebuild for oversized re-type batches.

### 5. Supporting correctness fixes
- `isIRI`/`isURI` returns `false` for skolemized blank nodes.
- `COUNT(*)` over `p1/p2+` sequence paths counts join multiplicity (not distinct pairs).

## Correctness model

Every fast path **defers** (`Ok(None)` → general pipeline / cursor merge) when it can't apply: overlay/time-travel below the indexed head, absent or novelty-only predicates, non-root policy, multi-ledger, history mode, overlay-flake translation failures, or a 0 result. Correctness never depends on a fast path firing — the general path remains the source of truth.

## Testing
- `it_query_sparql_indexed.rs` (+2,880): parity tests for each HEAD fast path, plus overlay/novelty and parallel variants (>50k-row datasets confirm the parallel paths actually fire; novelty tests cover asserts, retracts, and time-travel).
- `it_indexing_stats.rs` (+801): #1266 and ref-class re-attribution regressions (retract, re-type, oversized-batch rebuild).
- `it_query_seq_path_count_repro.rs`, `it_import_class_stats.rs`.
- `cargo fmt --all --check` + `clippy --all-features` clean; full-workspace `nextest` green except known-environmental suites (external-catalog and container-based integration tests) and a pre-existing flaky test (see follow-ups).

## Follow-ups (filed)
- **#1268** — parallel paths block the calling async worker under concurrent load (`spawn_blocking` + `'static` refactor).
- **#1269** — extend overlay-aware fast paths to the remaining aggregate shapes (GROUP BY top-K, COUNT DISTINCT, scalar SUM/AVG/MIN/MAX).
- **#1270** — class-to-class ref-edge `COUNT(*)` fold (now-exact `ref_classes`).
- **#1271** — `it_query_rdfs_class_repro` flakiness under concurrent nextest + bulk-import staging cleanup.
- **#1272** — incremental class-stat base-store-load-failure correctness gap (`IncrementalAbort` → rebuild).

### CODE REVIEW

File: fluree-db-query/src/fast_path_common.rs [hunk 1]
```diff
@@ -1702,6 +2125,259 @@ impl CursorSubjectCountStream {
...
+        let lo = leaves[i].first_key.s_id.as_u64();
+        let hi = if i + 1 < n {
+            leaves[i + 1].first_key.s_id.as_u64()
+        } else {
+            u64::MAX
+        };
+        // Ops with s_id in [lo, hi). Ops are sorted by s_id; advance the cursor.
+        while op_idx < ops.len() && ops[op_idx].s_id < lo {
+            op_idx += 1;
+        }
```
Comment (aaj3f) [inline fluree-db-query/src/fast_path_common.rs:2327]: I may be understanding this wrong, but it seems this would drop count consideration for overlay ops that are below the first index leaf's first subject.

It seems we'd need something like

```rust
// The first leaf's range starts at 0 so novelty ops below the predicate's
// first indexed subject (a low-id subject newly gaining this predicate) are
// not dropped — matching parallel_partition_count's bounds[0] == 0.
let lo = if i == 0 { 0 } else { leaves[i].first_key.s_id.as_u64() };
```

Maybe then also worth test / regression tests where a novelty assert at a subject id is below the indexed predicate's current minimum to ensure the count case isn't undercounting

File: fluree-db-indexer/src/build/incremental.rs [hunk 2]
```diff
@@ -2122,6 +2231,328 @@ pub async fn incremental_index(
...
+                                    Ok(m) => m,
+                                    Err(e) => {
+                                        tracing::warn!(
+                                            g_id, error = %e,
+                                            "Phase 3b: base subject-property scan failed; \
+                                             re-typed subjects may keep stale class stats"
+                                        );
+                                        continue;
+                                    }
+                                };
```
Comment (aaj3f) [inline fluree-db-indexer/src/build/incremental.rs:2295]: I'm trying to grok the degree to which this warn & continue should be acceptable. It seems maybe, given the PR description / commit promises etc that we should not simply warn but return `IndexerError::IncrementalAbort`?

There are other places where we swallow an error `batched_lookup_inbound_refs` and `batched_lookup_predicate_refs` but they seem to serve `ref_classes` which, as far as I can tell, aren't consumed meaningfully in the same way as this path is. But consistency-wise, perhaps they should go through the same "abort & rebuild" path as well?

**Review by aaj3f — APPROVED** [review]
It seems there are a few places where we have an overlay-parallel wrapper (`sum_star_join_overlay_parallel`, `sum_optional_join_overlay_parallel`, `try_modifier_intersect_overlay_parallel`, `try_union_constraint_overlay_parallel`) that calls `collect_resolved_overlay_ops` for every predicate and which walks all novelty flakes via `for_each_overlay_flake` before then dispatching `parallel_partition_count`. It's inside of that where the `total_rows < 50k || ncpu < 2` gate lives. For scenarios under that gate, then, the novelty collection is useless overhead, if I'm understanding correctly. It's also then re-done by the serial cursor-merge fallback. This seems like a double-effort perf scenario and a probably regression for the same queries before this PR

A cheap pre-gate check w/ a computed `total_rows` and a check like `available_parallelism() >= 2 && total_rows >= PARALLEL_STAR_MIN_ROWS`, if done before the `collect_resolved_overlay_ops` loop may solve for this

We may want some more test coverage for the claims (to validate them and avoid regressions):
  - The claim re: "parallel path fires" isn't tested. Given `parallel_partition_count` is serial when `ncpu < 2`, the parallel partition-boundary code would never be exercised
  - I don't see any tests that involve policy-hidden subjects such that there's a fold vs defer and where fold would over-count
  - Per the other comment I made re: novelty ids < index leaf ids, we likely want a test where overlay-delta novelty uses a low-id-subject < the leftmost index leaf's idea of "first"
  - We may want a test involving retraction where we retract an entire boundary leaf to zero etc to cover `merged < base`

Some additional in-line comments below but approving for whenever you're ready

File: fluree-db-query/src/fast_path_common.rs (hunk 1 above)
Comment (bplatz) [reply fluree-db-query/src/fast_path_common.rs:2327]: Fixed in dfd48b3b75d7b68718505041349ae2cdfac1845b: leaf 0 now ranges from 0, so a low-id existing subject newly gaining the predicate is counted (+ regression test).

File: fluree-db-indexer/src/build/incremental.rs (hunk 2 above)
Comment (bplatz) [reply fluree-db-indexer/src/build/incremental.rs:2295]: Fixed in dfd48b3b75d7b68718505041349ae2cdfac1845b: all four Phase 3b base-scan failures now return IncrementalAbort (rebuild) instead of warn-and-continue.
